### PR TITLE
[#1003850] add text masks

### PR DESCRIPTION
**Pull request by xaylpq**

### CODE REVIEW

File: localization/localization.n3 [hunk 1]
```diff
@@ -417,10 +417,36 @@
...
 
+<http://comindware.com/text/core/form/editors/textEditor#carNumberErrorMessageExpression> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy "Anatoly Vasilkov"; l10n:modifiedBy "Anatoly Vasilkov"; l10n:creationDate "2021-11-23T13:39:35.4288784Z"; l10n:modificationDate "2021-11-23T13:40:01.0209274Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "Enter car number in the format А999АА 099"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
+
+<http://comindware.com/text/core/form/editors/textEditor#carNumberPlaceholder> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy "Anatoly Vasilkov"; l10n:modifiedBy "Anatoly Vasilkov"; l10n:creationDate "2021-11-23T13:37:55.0466871Z"; l10n:modificationDate "2021-11-23T13:38:52.2767941Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "> A999AA 099"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
```
Comment (stasguryev) [inline localization/localization.n3:422]: пробел

File: resources/styles/theme.css [hunk 2]
```diff
@@ -270,6 +270,29 @@ select {
...
+    top: 20px;
+    right: 10px;
+    max-width: 180px;
+    background-color: var(--theme-colors__main);
+    border: 1px solid var(--border-color);
+    border-radius: 4px;
+    display: none;
+    padding: 5px;
+    text-align: center;
+    z-index: 1;
```
Comment (stasguryev) [inline resources/styles/theme.css]: ?

File: resources/styles/theme.css [hunk 3]
```diff
@@ -270,6 +270,29 @@ select {
...
+    transition: left 0.3s;
+    color: var(--theme-colors__error);
+}
+
+.error-message {
+    position: absolute;
+    top: 20px;
+    right: 10px;
+    max-width: 180px;
+    background-color: var(--theme-colors__main);
```
Comment (stasguryev) [inline resources/styles/theme.css]: + color

File: src/form/editors/TextEditorView.js [hunk 4]
```diff
@@ -13,17 +13,38 @@ const changeMode = {
...
     changeMode: 'blur',
     maxLength: undefined,
+    placeholderChar: '\u2000',
     mask: undefined,
-    placeholderChar: '_',
-    maskOptions: {},
     showTitle: true,
     allowEmptyValue: true,
     class: undefined,
     format: 'text',
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js]: change?

File: src/form/editors/TextEditorView.js [hunk 5]
```diff
@@ -48,31 +69,35 @@ const defaultOptions = () => ({
...
-            this.__setMaskByFormat(options.format, defOps);
-        } else {
-            this.mask = this.options.mask;
-        }
-
         this.__applyOptions(options, defOps);
         this.__debounceOnClearClick = _.debounce((...args) => this.__onClearClick(...args), DOUBLECLICK_DELAY);
+
+        this.format = options.format;
+        this.isMaskFormat = Object.values(serverMaskTypes).includes(options.format);
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js:59]: this.isMaskFormat Bolean(format)

File: src/form/editors/TextEditorView.js [hunk 6]
```diff
@@ -139,8 +174,19 @@ export default formRepository.editors.Text = BaseEditorView.extend({
     },
 
     __value(value, updateUi, triggerChange) {
+        if (this.isMaskFormat) {
+            const result = this.options.validationRegex.test(value?.trim());
+            if (value && !result) {
+                this.ui.errorIcon.css('display', 'inline');
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js]: this.el.classList.add

File: src/form/editors/TextEditorView.js [hunk 7]
```diff
@@ -164,6 +210,13 @@ export default formRepository.editors.Text = BaseEditorView.extend({
             this.__updateInputValue(value);
         }
     },
+    __getErrorMessage() {
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js]: __getErrorMessage + __resetErrorMessage = toggleErrorMessage

File: src/form/editors/TextEditorView.js [hunk 8]
```diff
@@ -188,31 +241,52 @@ export default formRepository.editors.Text = BaseEditorView.extend({
...
-        let additionalValidator;
-        switch (format) {
-            case 'email':
-                this.mask = emailMask;
-                additionalValidator = 'email';
-                defaults.emptyPlaceholder = Localizer.get('CORE.FORM.EDITORS.TEXTEDITOR.EMAILPLACEHOLDER');
+    __setMaskByFormat() {
+        switch (this.format) {
+            case serverMaskTypes.TEL:
+                this.options.validationRegex = /^((8|\+7)[\- ]?)?(\(?\d{3}\)?[\- ]?)?[\d\- ]{7,10}$/;
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js]: defaultValidators

File: src/form/editors/TextEditorView.js [hunk 9]
```diff
@@ -188,31 +241,52 @@ export default formRepository.editors.Text = BaseEditorView.extend({
...
+                this.options.mask = '+7 (999) 999-99-99';
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js]: !

File: src/form/editors/templates/textEditor.hbs [hunk 10]
```diff
@@ -1 +1,3 @@
+<span class="js-error-icon error-icon"><i class="fa fa-exclamation-circle"></i></span>
```
Comment (stasguryev) [inline src/form/editors/templates/textEditor.hbs]: Iconprefixer

File: demo/app/DemoConfig.js [hunk 11]
```diff
@@ -44,6 +44,11 @@ export default {
                     displayName: 'Form using form behaviour',
```
Comment (stasguryev) [inline demo/app/DemoConfig.js:44]: screenShots

File: resources/styles/theme.css (hunk 2 above)
Comment (xaylpq) [reply resources/styles/theme.css]: image

File: demo/app/DemoConfig.js (hunk 11 above)
Comment (xaylpq) [reply demo/app/DemoConfig.js:44]: image

File: resources/styles/theme.css (hunk 2 above)
Comment (xaylpq) [reply resources/styles/theme.css]: * Deleted

File: resources/styles/theme.css (hunk 3 above)
Comment (xaylpq) [reply resources/styles/theme.css]: * Deleted

File: src/form/editors/TextEditorView.js (hunk 4 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js]: Done

File: src/form/editors/TextEditorView.js (hunk 5 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js:59]: ?

File: src/form/editors/TextEditorView.js (hunk 6 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js]: * Deleted

File: src/form/editors/TextEditorView.js (hunk 7 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js]: * Deleted

File: src/form/editors/TextEditorView.js (hunk 8 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js]: Done

File: src/form/editors/TextEditorView.js (hunk 9 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js]: Done

File: src/form/editors/templates/textEditor.hbs (hunk 10 above)
Comment (xaylpq) [reply src/form/editors/templates/textEditor.hbs]: * Deleted

File: localization/localization.n3 [hunk 12]
```diff
@@ -419,7 +419,47 @@
...
 
-<http://comindware.com/text/core/form/editors/textEditor#emailPlaceholder> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy ""; l10n:modifiedBy ""; l10n:creationDate "2018-09-29T16:30:04.2640482Z"; l10n:modificationDate "2018-09-29T16:30:07.0212267Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "Enter email"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
+<http://comindware.com/text/core/form/editors/textEditor#emailErrorMessageExpression> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy "Anatoly Vasilkov"; l10n:modifiedBy "Anatoly Vasilkov"; l10n:creationDate "2021-11-23T13:26:47.6713548Z"; l10n:modificationDate "2021-11-23T13:27:04.1233877Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "Enter e-mail in the format *@*.*"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
```
Comment (stasguryev) [inline localization/localization.n3]: localiztion.josn ?

File: src/form/editors/TextEditorView.js [hunk 13]
```diff
@@ -90,21 +98,20 @@ export default formRepository.editors.Text = BaseEditorView.extend({
...
+                    Inputmask({
+                        regex: this.options.regexMask
+                    }).mask(this.ui.input[0]);
+                    break;
+            }
         }
     },
 
-    onDestroy() {
-        this.maskedInputController && this.maskedInputController.destroy();
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js:105]: Inputmask.remove?

File: src/form/editors/TextEditorView.js [hunk 14]
```diff
@@ -139,14 +146,10 @@ export default formRepository.editors.Text = BaseEditorView.extend({
...
-        let realValue = value;
-        if (!updateUi && value && this.options.format === 'tel') {
-            realValue = realValue.replace(/[^\d]/g, '');
-        }
-        if (this.value === realValue) {
+        if (this.value === value) {
             return;
         }
-        this.value = realValue;
+        this.value = value;
```
Comment (stasguryev) [inline src/form/editors/TextEditorView.js:156]: unmasked value?

File: localization/localization.n3 (hunk 12 above)
Comment (arterm-sedov) [reply localization/localization.n3]: Enter the e-mail address in *@*.* format
Введите адрес эл. почты в формате *@*.*

File: localization/localization.n3 [hunk 15]
```diff
@@ -419,7 +419,47 @@
...
+
+<http://comindware.com/text/core/form/editors/textEditor#emailRegexp> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy "Anatoly Vasilkov"; l10n:modifiedBy "Anatoly Vasilkov"; l10n:creationDate "2021-11-25T15:20:32.9275443Z"; l10n:modificationDate "2021-11-25T15:20:34.4695473Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "^(([a-zа-я0-9_-]+\\.)*[a-zа-я0-9_-]+@[a-zа-я0-9-]+(\\.[a-zа-я0-9-]+)*\\.[a-zа-я]{2,6})?$"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
+
+<http://comindware.com/text/core/form/editors/textEditor#indexMessageExpression> a l10n:Resource; l10n:descr ""; l10n:isLocalizable true; l10n:needsReview true; l10n:createdBy "Anatoly Vasilkov"; l10n:modifiedBy "Anatoly Vasilkov"; l10n:creationDate "2021-11-23T13:51:01.6341809Z"; l10n:modificationDate "2021-11-23T13:51:48.2282984Z"; l10n:text [a l10n:Text; l10n:lang l10n:en; l10n:data "Enter index in the format 000000"; l10n:originalData "{62D69EB8-800C-480F-9571-E11F3DF477C0}"].
```
Comment (arterm-sedov) [inline localization/localization.n3]: Enter the postal code in 000000 format
Введите индекс в формате 000000

File: src/form/editors/TextEditorView.js (hunk 13 above)
Comment (xaylpq) [reply src/form/editors/TextEditorView.js:105]: Yes

File: localization/localization.n3 (hunk 15 above)
Comment (xaylpq) [reply localization/localization.n3]: Done

File: localization/localization.n3 (hunk 12 above)
Comment (xaylpq) [reply localization/localization.n3]: Done

**Comment by xaylpq** [comment]
>

**Pull request merged**